src/client/api/metrics.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * (C) Copyright 2020-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/utsname.h>
+#include <daos/common.h>
+#include <daos/job.h>
+#include <daos/tls.h>
+#include <gurt/telemetry_common.h>
+#include <gurt/telemetry_consumer.h>
+#include <gurt/telemetry_producer.h>
+
+#define INIT_JOB_NUM		1024
+bool client_metric;
+bool client_metric_retain;
+
+#define MAX_IDS_SIZE(num) (num * D_TM_METRIC_SIZE)
+/* The client side metrics structure looks like
+ * root/job_id/pid/....
+ */
+int
+dc_tm_init(void)
+{
+	struct d_tm_node_t *job_node;
+	struct d_tm_context *current_ctx;
+	struct daos_thread_local_storage *dtls;
+	int		    metrics_tag;
+	pid_t		    pid;
+	int		    rc;
+
+	d_getenv_bool(DAOS_CLIENT_METRICS_ENV, &client_metric);
+	if (!client_metric)
+		return 0;
+
+	d_getenv_bool(DAOS_CLIENT_METRICS_RETAIN_ENV, &client_metric_retain);
+
+	metrics_tag = D_TM_CLIENT_PROCESS | D_TM_OPEN_OR_CREATE;
+	if (client_metric_retain)
+		metrics_tag |= D_TM_RETAIN_SHMEM;
+	else
+		metrics_tag |= D_TM_RETAIN_SHMEM_IF_NON_EMPTY;
+
+	rc = d_tm_init(DC_TM_JOB_ROOT_ID, MAX_IDS_SIZE(INIT_JOB_NUM), metrics_tag);
+	if (rc != 0) {
+		D_ERROR("init job root id %u: %d\n", DC_TM_JOB_ROOT_ID, rc);
+		return rc;
+	}
+
+	pid = getpid();
+	D_INFO("INIT %s/%u metrics\n", dc_jobid, pid);
+	rc = d_tm_add_metric(&job_node, D_TM_DIRECTORY,
+			     "job id directory", "dir",
+			     "%s/%u", dc_jobid, pid);
+	/* Close job root sheme */
+	d_tm_fini();
+	if (rc != 0) {
+		D_ERROR("add metric %s/%u failed: %d\n", dc_jobid, pid, rc);
+		D_GOTO(out, rc);
+	}
+
+	metrics_tag = D_TM_CLIENT_PROCESS;
+	if (client_metric_retain)
+		metrics_tag |= D_TM_RETAIN_SHMEM;
+	rc = d_tm_init(pid, MAX_IDS_SIZE(INIT_JOB_NUM), metrics_tag);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+	current_ctx = d_tm_open(pid);
+	if (current_ctx == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
```

Comment:
I am not 100% sure, but I'm pretty sure that this is the wrong way to do this. Instead of manually managing the shm segments, I think it would be better to use [d_tm_add_ephemeral_dir()](https://github.com/daos-stack/daos/blob/master/src/gurt/telemetry.c#L2455) to attach the child segments to the parent. This will make walking down from the root work better.

@kjacque added this feature to the telemetry for dynamic pool support a few years ago; she may have more thoughts about how to organize this.

Reply:
Ah, I did not see there is d_tm_add_ephemeral_dir(). hmm, this looks similar as this function. Yes, indeed, I should use this function. let me refresh the patch. Thanks!